### ISSUE TITLE
[Discussion]: New $ and $$ type inference makes code harder to read

### ISSUE THREAD

**Issue by ldiqual** [puppeteer/puppeteer#8579]
### Bug description

v15 tries to [infer types](https://github.com/puppeteer/puppeteer/pull/8547) for simple selectors like `'a'` or `'tr'`. It removed the ability to provide explicit types via generics and requires casting for non-trivial selectors.

I would argue that most code doesn't use trivial selectors and provides a nested path or classes instead. Developers often know the return type already and the new inference doesn't allow them to provide these types.

```typescript
// v14
await page.$<HTMLAnchorElement>('a')
await page.$<HTMLAnchorElement>('.class > a')

// v15
await page.$('a')
(await page.$('.class > a')) as ElementHandle<HTMLAnchorElement> | null
```

Simply put, while the simple cases are now shorter, the more complex cases are now longer/harder to read.

I propose that puppeteer reinstates the ability to specify a generic type *in addition to* the current smart inference.

### Puppeteer version

v15.1.1

### Node.js version

v16.15.1

### npm version

Yarn 3.2.1

### What operating system are you seeing the problem on?

macOS

**Comment by OrKoN** [comment]
Actually, I agree with this. Having both would be great and it looks like we overlooked the impact on the more complex selectors. @jrandolf WDYT? I am not sure if it's possible to have both ways supported.

**Comment by jrandolf-google** [comment]
Hey, thanks for bringing this up! While making the PR, I wanted to discuss the reasoning behind this decision in detail, but for brevity, I chose to just leave it out.

There are two core problems with allowing users to set their own return type (and one extra side note):
1. A malicious/unknowing user can set the return type to something completely wrong and get away with it. This is a result of the `ReturnType` *type* parameter not being used in any way for type inference. For example, the `v15` method uses the selector *argument* parameter to deduce the return type. 
2. Even if the user is correct, this doesn't imply they always will be. For example, in your example, suppose you just wrote `.class` instead of `.class > a`. At the moment **you** code this, you may know it's correct, but a year later, suppose **someone else** changed the tag name of some `.class` element. Then the code has potential to break at the runtime level while still being correct at the type level.
- [Side note] @[ahejlsberg](https://github.com/ahejlsberg), the lead architect of TypeScript, has also commented about this ([here](https://github.com/microsoft/TypeScript/pull/42067#issuecomment-754860862)) remarking this is an anti-pattern.

From these problems, you can deduce more problems such as maintainers (us) getting issues due to some library lying to their users about a runtime problem when it was in fact issue (2) in play.

Now all is not lost. TypeScript has some great features such as template types. For example, we could do the following:
```ts
function $<Selector extends keyof HTMLElementTagNameMap>(
    selector: `${string} ${Selector}`
): Promise<ElementHandle<HTMLElementTagNameMap[Selector]> | null>;
```
This would deduce your `.class > a` to an `HTMLAnchorElement`. As a side note, this would also strengthen issue (2) since any type you put in a hypothetical `ReturnType` **will now always have a chance to be wrong** since any selectors not conforming to `${string} ${Selector}` will be based on a non-tagname based one.

[EDIT] Tested the above, and it doesn't work, however there is a way to do it as seen in https://github.com/microsoft/TypeScript/issues/29037. Basically, you write a parser in TypeScript's type language and parse the selector. I think once this is settled, we can move toward such a selector, but for now, we will keep the selectors as they are.

**Comment by ldiqual** [comment]
Got it. I do agree that `page.$<HTMLAnchorElement>(...)` is essentially a cast in disguise. I guess the new approach is to place the decision to unsafe cast into the developer's hand, instead of hiding it. I was also under the assumption that complex selectors could never be inferred in the future, but [this](https://github.com/g-plane/typed-query-selector) is obviously proving me wrong.

My main concern is that in the meantime, this example:

```typescript
// Before
await page.$<HTMLAnchorElement>('.class > a')

// After
(await page.$('.class > a')) as ElementHandle<HTMLAnchorElement> | null
```

requires developers to look at the interface of `$`, write `ElementHandle`, provide the element type, add nullability, all while running into the risk of missing some of those.

I wonder if there's a middle ground where either:

 1. `$` would accept another function parameter – an element name – and fail if the handle doesn't have the proper type:

```typescript
// Validates that the selector returns an anchor, then returns it casted
const anchor = await page.$('.class > a', { element: 'a' })
```

 2. puppeteer would expose a function that validates, then casts an element handle.

```typescript
const handle = await page.$('.class > a')
const anchor = await validateHandle(handle, 'a')
```

But if you do have typed query selectors on the roadmap, feel free to ignore all of this, I'll cast in the meantime 👍

**Comment by jrandolf-google** [comment]
Thanks for understanding. We'll consider the proposed solutions; they are more feasible since they implement runtime checks, however I'm hard-pressed those solutions should belong in puppeteer. @OrKoN WDYT? We could implement something like
```ts
export async function isElementHandleOf<T extends keyof HTMLElementTagNameMap>(handle: ElementHandle, tagName: T): handle is ElementHandle<HTMLElementTagNameMap[T]> {
  return await handle.evaluate((e, tagName) => {
    return !!(e instanceof Element && e.tagName.match(new RegExp(tagName, "i")));
  }, tagName);
}
```

P.S. Don't use `ElementHandle` directly. Use `HandleFor` which automatically deduces the correct handle.
